src/database/firestore.ts

```diff
@@ -1,71 +1,162 @@
-        return Observable.create(observer => {
-            this.db.collection(collection).doc(docId).get()
-                .then((doc) => {
-                    if(doc.exists) {
-                        observer.next(doc.data());
-                        observer.complete();
-                    } else {
-                        observer.error('Document does not exist');
+    addOrUpdateDocById(collection: string, docId: string, entity: object): Observable<void>{
+            return from(this.db.collection(collection).doc(docId).set(entity));
+    }
+
+    getDoc(collection: string, docId: string): Observable<Document> {
+        return from(this.db.collection(collection).doc(docId).get())
+            .pipe(
+                map(docSnapshot => {
+                    if (docSnapshot.exists)
+                        return {
+                            id: docId,
+                            data: docSnapshot.data()
+                        };
+                    else
+                        throw new DocumentNotFoundError(docId);
+                })
+            )
+    }
+
+    getCollection(collection: string): Observable<Document[]> {
+        return from(this.db.collection(collection).get())
+            .pipe(
+                map(querySnapshot => {
+                    const docs = [];
+                    const queryDocumentSnapshots = querySnapshot.docs;
+                    for(let i = 0; i < queryDocumentSnapshots.length; i++) {
+                        const doc = queryDocumentSnapshots[i];
+                        docs.push({
+                            id: doc.id,
+                            data: doc.data()
+                        });
                     }
+                    return docs;
                 })
-                .catch((err) => {
-                    observer.error(err);
-                });
-        });
+            );
+    }
+
+    getUsers() {
+        return this.getCollection(USER_COLLECTION);
+    }
+
+    doesExist(collection: string, docId: string): Observable<boolean> {
+        return from(this.db.collection(collection).doc(docId).get())
+            .pipe(
+                map(docSnapshot => docSnapshot.exists)
+            );
     }
 
-    updateDoc(collection: string, docId: string, entity: object) {
-        return from(this.db.collection(collection).doc(docId).update(entity))
+    updateDoc(collection: string, docId: string, entity: object): Observable<void> {
+        return from(this.db.collection(collection).doc(docId).update(entity));
     }
 
-    deleteDoc(collection: string, docId: string) {
+    deleteDoc(collection: string, docId: string): Observable<void> {
         return from(this.db.collection(collection).doc(docId).delete());
     }
 
     closeConnection() {
-        return from(this.app.delete())
+        return from(this.app.delete());
+    }
+
+    private storeUserInfoOnLogin(): Observable<User> {
+        const attachExistence = firebaseUser =>
```

Comment:
What’s going on in this method

Reply:
StoreUserInfoOnLogin I mean

Reply:
A lot lol.
Basically it takes the Observable that you made for whenever the auth state changes, then it splits it into two Observables based on whether the person is logging in or logging out, and if the user is logging in, it checks whether that person exists in the database. If he/she doesn't, it stores that person's basic information in the db. If he/she does, we just update their profile picture. It then merges the two streams together into one Observable.